Helper/Cart.php

```diff
@@ -526,6 +526,56 @@ public function replicateQuoteData($source, $destination)
         $this->discountHelper->cloneAmastyGiftCards($source->getId(), $destination->getId());
     }
 
+    /**
+     * Reserve Order Id for the first immutable quote created.
+     * Store it in BoltReservedOrderId field in the parent quote
+     * as well as in any subsequently created immutable quote.
+     * Defer setting ReservedOrderId on the parent quote
+     * until the quote to order submission.
+     * Reason: Some 3rd party plugins read this value and if set
+     * consider the quote to be a complete order.
+     *
+     * @param Quote $immutableQuote
+     * @param Quote $quote
+     * @throws \Magento\Framework\Exception\AlreadyExistsException
+     */
+    protected function reserveOrderId($immutableQuote, $quote)
+    {
+        $reservedOrderId = $immutableQuote->getBoltReservedOrderId();
+        if (!$reservedOrderId) {
+            $reservedOrderId = $immutableQuote->reserveOrderId()->getReservedOrderId();
+            $immutableQuote->setBoltReservedOrderId($reservedOrderId);
+            $quote->setBoltReservedOrderId($reservedOrderId);
+            $this->quoteResourceSave($quote);
+        } else {
+            $immutableQuote->setReservedOrderId($reservedOrderId);
```

Comment:
I'm not sure if I follow here. what is this else block for?

Reply:
reserveOrderId() is called on the first immutable quote and then propagated to parent and to subsequent immutables in BoltReservedOrderId field and in the end copied from parent BoltReservedOrderId to ReservedOrderId. This else is for consistency, to keep every subsequent immutable having all the fields populated. Actually, one more important reason, some lines below $cart['display_id'] = $immutableQuote->getReservedOrderId() . ' / ' . $immutableQuote->getId(); It is expected to be there. So every call to getReservedOrderId remain unchanged.